include/dali/core/span.h

```diff
@@ -66,6 +66,13 @@ class span {
   span &operator=(const span &other) noexcept = default;
 
   // [span.sub], span subviews
+  template <index_type count>
+  DALI_HOST_DEV constexpr auto first() const { return span<element_type, count>(data_, count); }
+  DALI_HOST_DEV constexpr auto first(index_type count) const { return span<element_type, dynamic_extent>(data_, count); }
+
+  template <index_type count>
+  DALI_HOST_DEV constexpr auto last() const { return span<element_type, count>(data_ + size() - count, count); }
+  DALI_HOST_DEV constexpr auto last(index_type count) const { return span<element_type, dynamic_extent>(data_ + size() - count, count); }
```

Comment:
C++14 `auto` return type is not compatible with SFINAE (it causes a hard error), because it requires instantiation of the function in order to even check if substitution is possible.
`span` is a _very_ common utility and it's likely that functions like first/last would be used that way.
```suggestion
  DALI_HOST_DEV constexpr span<element_type, count> first() const { return { data_, count }; }
  DALI_HOST_DEV constexpr span<element_type> first(index_type count) const { return { data_, count }; }

  template <index_type count>
  DALI_HOST_DEV constexpr span<element_type, count> last() const { return { data_ + size() - count, count }; }
  DALI_HOST_DEV constexpr span<element_type> last(index_type count) const { return { data_ + size() - count, count }; }
```

Reply:
Sure, will revert to that.